### PR TITLE
new lib to manage duplicates

### PR DESCRIPTION
**Pull request by AlicVB**
- show a list of all available duplicate in darkroom with thumbnails
- add btn to create classic or "virgin" duplicates of the current image
- click and stay pressed to quickly show the duplicate without switching to the image (easy compare)
- dble-click switch to the duplicate

The lib can be seen of a film-strip "duplicate" but film-strip with grouping is not really handy to manage duplicate, as groups often contains other images (jpeg, bracketing, ...) Plus it contain some specific features.

Some points to review : 
- is it safe to use "caption" field in image table to store duplicate name/description ? It doesn't seem to be use elsewhere in the code.
- if we decide to include this lib, I'll better create a new global signal to switch to a new picture. (currently, I reuse the filmstrip fct)

### CODE REVIEW

File: src/libs/duplicate.c [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+static void _lib_duplicate_delete(GtkButton *button, dt_lib_module_t *self)
+{
+  if(dt_conf_get_bool("ask_before_remove"))
+  {
+    GtkWidget *dialog;
+    GtkWidget *win = dt_ui_main_window(darktable.gui->ui);
+
+    dialog = gtk_message_dialog_new(GTK_WINDOW(win), GTK_DIALOG_DESTROY_WITH_PARENT, GTK_MESSAGE_QUESTION,
+                                    GTK_BUTTONS_YES_NO,
+                                    "do you really want to remove 1 image from the collection?");
```
Comment (LebedevRI) [inline src/libs/duplicate.c]: This also probably needs `_()`

Comment (AlicVB) [inline src/libs/duplicate.c]: yep... fixed :)

**Comment by TurboGit** [comment]
First of all great to see this. I have just tested it and found this quite nice, a very good idea filling a missing feature!

Not sure for caption but looks ok to me.

I would certainly put the duplicate module under the history. I don't think it is ok to separate the history and snapshot as both libs are really to be used together (comparing images at different stage).

In the UI side I would put some space horizontally between the thumb the caption and the number. BTW, is the number bringing any useful information here? No strong opinion but I would not put it here. Instead of the number I would put a small trash to remove the duplicate.

**Comment by TurboGit** [comment]
Also I would not display the stars on the selected thumb. Small, difficult to read and cannot be changed... So not really useful to me.

**Comment by TurboGit** [comment]
The 'virgin' duplicate creation does not even have the base-curve not the sharpen module. Is that expected for RAW image? That's not the way it works from the lighttable and so found this a bit misleading.

**Comment by AlicVB** [comment]
thks for the comments.
- 'virgin' duplicates now include auto-applied history, like every non-developped image
- added a remove btn
- put the lib just under history
For the other points : 
- the duplicate nb is useful to have a ref, if you don't set any captions
- about the thumbnail display, it's a limitation of the actual api. it's quite easy to change, but will require some thinking about what options to expose. I think it's better to postpone this, and let this code independent from the rest of dt

**Comment by houz** [comment]
No, using caption is not ok. While it's one of the old, now unused, fields I wouldn't assume that nothing is in there. If it's something the user can enter then maybe just use `Xmp.dc.title`?

File: src/libs/duplicate.c [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+}
+
+static void _lib_duplicate_init_callback(gpointer instance, dt_lib_module_t *self);
+
+static gboolean _lib_duplicate_caption_out_callback(GtkWidget *widget, GdkEvent *event, dt_lib_module_t *self)
+{
+  int imgid = GPOINTER_TO_INT(g_object_get_data(G_OBJECT(widget),"imgid"));
+  
+  sqlite3_stmt *stmt;
+  // we write the content of the texbox to the caption field
```
Comment (houz) [inline src/libs/duplicate.c]: tex_t_box

File: src/libs/duplicate.c [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+static void _lib_duplicate_delete(GtkButton *button, dt_lib_module_t *self)
+{
+  if(dt_conf_get_bool("ask_before_remove"))
+  {
+    GtkWidget *dialog;
+    GtkWidget *win = dt_ui_main_window(darktable.gui->ui);
+
+    dialog = gtk_message_dialog_new(GTK_WINDOW(win), GTK_DIALOG_DESTROY_WITH_PARENT, GTK_MESSAGE_QUESTION,
+                                    GTK_BUTTONS_YES_NO,
+                                    _("do you really want to remove 1 image from the collection?"));
```
Comment (houz) [inline src/libs/duplicate.c]: Better reuse the string from elsewhere so translations stay in sync.

File: src/libs/duplicate.c [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+    else if(event->type == GDK_2BUTTON_PRESS)
+    {
+      // to select the duplicate, we reuse the filmstrip proxy
+      dt_view_filmstrip_scroll_to_image(darktable.view_manager,imgid,TRUE);
+    }
+  }
+  else if(event->button == 3)
+  {
+    // context-menu
+    GtkMenuShell *menu = GTK_MENU_SHELL(gtk_menu_new());
```
Comment (houz) [inline src/libs/duplicate.c]: In general darktable doesn't use context menus. Just saying.

File: src/libs/duplicate.c [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+  gtk_box_pack_start(GTK_BOX(self->widget), hb, FALSE, FALSE, 0);
+  gtk_container_add(GTK_CONTAINER(sw), d->duplicate_box);
+  gtk_box_pack_start(GTK_BOX(self->widget), sw, FALSE, FALSE, 0);  
+
+  gtk_widget_show_all(self->widget);
+  
+  dt_control_signal_connect(darktable.signals, DT_SIGNAL_DEVELOP_IMAGE_CHANGED, G_CALLBACK(_lib_duplicate_init_callback), self);
+  dt_control_signal_connect(darktable.signals, DT_SIGNAL_DEVELOP_INITIALIZE, G_CALLBACK(_lib_duplicate_init_callback), self);
+  dt_control_signal_connect(darktable.signals, DT_SIGNAL_DEVELOP_MIPMAP_UPDATED, G_CALLBACK(_lib_duplicate_mipmap_updated_callback), (gpointer)self);
+}
```
Comment (houz) [inline src/libs/duplicate.c]: You probably also want to listen to `DT_SIGNAL_COLLECTION_CHANGED` or whatever gets raised when a new image gets loaded to update your list.

**Comment by AlicVB** [comment]
thks for your comments, they should be fixed now.

**Comment by Iambecomeroot** [comment]
Any updates on this?

**Comment by TurboGit** [comment]
@AlicVB Thanks, I have merged this and ported to recent darktable API for this lib to compile. A very nice addition for 2.6.

**Pull request closed without merging**